File: maximal_subarray.cpp
// Time complexity-O(n)
// space complexity O(1);
// Did this code successfully run on Leetcode :yes
// Any problem you faced while coding this : no

// at each position  decide if the element should be added to the window or not

#include<algorithm>
#include<iostream>
#include<vector>
using namespace std;

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int max_sum=nums[0];int curr_sum=nums[0];
        int start=0 ;int end=1;
        for(int i=1; i<nums.size();i++){
            
            if(nums[i] > curr_sum+nums[i]){
                start=i;
            }
            
            
            curr_sum = max(nums[i]+curr_sum,nums[i]);
            
            if(curr_sum > max_sum){
                end++;
            }
            max_sum =max(curr_sum,max_sum);
            
        }
        cout<<start<<endl;
        cout<<end<<endl;
       return max_sum; 
    }
    
};

File: maximal_subarray.py
# Time Complexity : O(n) because we are iterating through the array
# Space Complexity : O(1) because we are not using any extra space
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach in three sentences only
"""
We iterate through the array and keep the increasing the size of the subarray until we find that that subarray sum is less than 0
"""

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        maxsum = subarray = nums[0]

        for i in range(1,len(nums)):
            subarray = max(nums[i], subarray + nums[i])
            maxsum = max(maxsum, subarray)

        return maxsum 